iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift

```diff
@@ -254,13 +266,46 @@ class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
         return CGPoint(x: proposedContentOffset.x, y: proposedContentOffset.y + self.scrollOffset)
     }
 
+    /// NOTE: Disappearing does not mean that the item will not be visible after the animation.
+    /// Per the docs:  "For each element on screen before the invalidation, finalLayoutAttributesForDisappearingXXX will be called..."
     override func finalLayoutAttributesForDisappearingItem(at itemIndexPath: IndexPath)
     -> UICollectionViewLayoutAttributes? {
-        // Items that are just moving are marked as "deleted" by the collection view.
+
+        // Remember which items were visible before the animation started so we don't attempt to modify
+        // their animations later.
+        self.indexPathsVisibleBeforeAnimation.insert(itemIndexPath)
+
+        // Items that are just moving are marked as "disappearing"" by the collection view.
         // Only animate changes to items that are actually being deleted otherwise weird animation issues
         // will arise.
         guard self.deletedIndexPaths.contains(itemIndexPath) else { return nil }
 
         return super.finalLayoutAttributesForDisappearingItem(at: itemIndexPath)
     }
+
+    /// NOTE: "Appearing" does not mean the item wasn't visible before the animation.
+    /// Per the docs: "For each element on screen after the invalidation, initialLayoutAttributesForAppearingXXX will be called..."
+    override func initialLayoutAttributesForAppearingItem(at itemIndexPath: IndexPath)
+    -> UICollectionViewLayoutAttributes? {
+
+        let attributes = super.initialLayoutAttributesForAppearingItem(at: itemIndexPath)
+
+        // Determine if this item existed before, but was not visible. If so, we need
+        // to modify it's attributes to make it appear properly.
+        if !self.indexPathsVisibleBeforeAnimation.contains(itemIndexPath),
+           !self.insertedIndexPaths.contains(itemIndexPath) {
+            
+            var normalizedZOffset = self.getNormalizedZOffsetForItem(at: itemIndexPath,
+                                                                     givenZPosition: self.initialZPosition)
+            normalizedZOffset = clamp(normalizedZOffset, -1, 1)
+            let modifiedAttributes = self.layoutAttributesForItemAt(indexPath: itemIndexPath,
+                                                                    withNormalizedZOffset: normalizedZOffset)
+
+            modifiedAttributes?.center.y += self.initialZPosition - self.zPosition
+
+            return modifiedAttributes
+        }
```

Comment:
Items that aren't visible before the insert/delete, but become visible afterwards, act weird because they're animating while the collection view is scrolling. This is here to compensate for that.